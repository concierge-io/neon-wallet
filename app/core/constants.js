// @flow

export const NEON_WALLET_RELEASE_LINK = 'https://github.com/CityOfZion/neon-wallet/releases'

export const NETWORK = {
  MAIN: 'MainNet',
  TEST: 'TestNet',
  APHELION_MAIN: 'MainNet-A',
  APHELION_TEST: 'TestNet-A'
}

export const EXPLORER = {
  NEO_TRACKER: 'Neotracker',
  NEO_SCAN: 'Neoscan',
  ANT_CHAIN: 'Antchain'
}

export const ASSETS = {
  NEO: 'NEO',
  GAS: 'GAS'
}

export const ASSET_ID = {
  NEO: 'c56f33fc6ecfcd0c225c4ab356fee59390af8560be0e930faebe74a6daff7c9b',
  GAS: '602c79718b16e442de58778e148d0b1084e3b2dffd5de6b7b16cee7969282de7'
}

export const ASSETS_LABELS = {
  [ASSETS.NEO]: 'Neo',
  [ASSETS.GAS]: 'Gas'
}

export const DEFAULT_REQ = {
  jsonrpc: '',
  method: '',
  params: [],
  id: 0
}

export const ROUTES = {
  HOME: '',
  DASHBOARD: '/dashboard',
  CREATE_WALLET: '/create',
  ENCRYPT_KEY: '/encrypt-key',
  LOGIN_PRIVATE_KEY: '/login-private-key',
  LOGIN_LOCAL_STORAGE: '/login-local-storage',
  LOGIN_LEDGER_NANO_S: '/login-ledger-nano-s',
  LOGIN_NEP2: '/login-encrypted',
  LOGIN_TOKEN_SALE: '/login-token-sale',
  TOKEN_SALE: '/token-sale',
  SETTINGS: '/settings'
}

export const NOTIFICATION_LEVELS = {
  ERROR: 'error',
  SUCCESS: 'success',
  INFO: 'info',
  WARNING: 'warning'
}

export const NOTIFICATION_POSITIONS = {
  TOP_CENTER: 'tc',
  TOP_RIGHT: 'tr',
  TOP_LEFT: 'tl',
  BOTTOM_CENTER: 'bc',
  BOTTOM_RIGHT: 'br',
  BOTTOM_LEFT: 'bl'
}

export const BIP44_PATH =
  '8000002C' +
  '80000378' +
  '80000000' +
  '00000000' +
  '00000000'

export const MODAL_TYPES = {
  SEND_TRANSACTION: 'SEND_TRANSACTION',
  CONFIRM: 'CONFIRM'
}
