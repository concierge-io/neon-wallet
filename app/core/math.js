// @flow

// https://stackoverflow.com/questions/4912788/truncate-not-round-off-decimal-numbers-in-javascript
export const truncateNumber = (num: number, places: number): number =>
  Math.trunc(num * Math.pow(10, places)) / Math.pow(10, places)
