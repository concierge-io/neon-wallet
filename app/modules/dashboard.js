// @flow
import { LOGOUT } from './account'

// Constants
export const TOGGLE_SEND_PANE = 'TOGGLE_SEND_PANE'

// Actions
export function togglePane (pane: string) {
  return {
    type: TOGGLE_SEND_PANE,
    payload: { pane }
  }
}

// state getters
export const getSendPane = (state) => state.dashboard.sendPane
export const getConfirmPane = (state) => state.dashboard.confirmPane

const initialState = {
  sendPane: true,
  confirmPane: true
}

export default (state: Object = initialState, action: Object) => {
  switch (action.type) {
    case TOGGLE_SEND_PANE:
      const { pane } = action.payload
      let newState = {}
      newState[pane] = !state[pane]
      return {
        ...state,
        ...newState
      }
    case LOGOUT:
      return initialState
    default:
      return state
  }
}
