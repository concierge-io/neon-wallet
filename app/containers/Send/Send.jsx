// @flow
import React, { Component } from 'react'
import SplitPane from 'react-split-pane'
import ReactTooltip from 'react-tooltip'
import { validateTransactionBeforeSending } from '../../core/wallet'
import { ASSETS_LABELS, ASSETS } from '../../core/constants'

type Props = {
  togglePane: Function,
  showErrorNotification: Function,
  sendTransaction: Function,
  toggleAsset: Function,
  neo: number,
  gas: number,
  nep5: string[],
  tokens: Object[],
  balances: Object[],
  confirmPane: boolean,
  selectedAsset: string,
  selectedHash: string
}

type State = {
  sendAmount: string,
  sendAddress: string,
}

export default class Send extends Component<Props, State> {
  state = {
    sendAmount: '',
    sendAddress: ''
  }

  confirmButton: ?HTMLButtonElement

  // open confirm pane and validate fields
  openAndValidate = () => {
    const { neo, gas, selectedAsset, selectedHash, togglePane, showErrorNotification, balances } = this.props
    const { sendAddress, sendAmount } = this.state
    const { error, valid } = validateTransactionBeforeSending(neo, gas, selectedAsset, selectedHash, balances, sendAddress, sendAmount)
    if (valid) {
      togglePane('confirmPane')
    } else {
      showErrorNotification({ message: error })
    }
  }

  nextAsset = () => {
    const { selectedAsset, selectedHash, nep5, tokens, toggleAsset } = this.props
    let newAsset = ASSETS_LABELS.NEO;
    let currentIndex = nep5.findIndex( hash => hash === selectedHash );
    let nextHash = nep5[currentIndex + 1];
    let nextSymbol = tokens[nextHash] ? tokens[nextHash].symbol : undefined;
    if( selectedHash &&  nextHash && selectedAsset !== ASSETS_LABELS.NEO){
      newAsset = nextHash;
    }else if(selectedAsset === ASSETS_LABELS.NEO){
      newAsset = ASSETS_LABELS.GAS;
      nextSymbol = ASSETS_LABELS.GAS;
    }else if(selectedAsset === ASSETS_LABELS.GAS && Object.keys(tokens).length > 0){
      newAsset = tokens[ Object.keys(tokens)[0] ].symbol
    }
    if(newAsset === ASSETS_LABELS.NEO){
      nextSymbol = ASSETS_LABELS.NEO;
    }
    toggleAsset(nextSymbol ? nextSymbol : (currentIndex + 2)+'Â°', newAsset); //add +2 since humans count from 1
  }

  // perform send transaction
  sendTransaction = () => {
    const { sendTransaction, togglePane } = this.props
    const { sendAddress, sendAmount } = this.state
    sendTransaction(sendAddress, sendAmount).then(() => {
      togglePane('confirmPane')
      this.resetForm()
    })
  }

  resetForm () {
    this.setState({
      sendAddress: '',
      sendAmount: ''
    })
    if (this.confirmButton) {
      this.confirmButton.blur()
    }
  }

  render () {
    const { confirmPane, selectedAsset } = this.props
    const { sendAddress, sendAmount } = this.state
    const confirmPaneClosed = confirmPane ? '100%' : '69%'

    return (
      <SplitPane className='confirmSplit' split='horizontal' size={confirmPaneClosed} allowResize={false}>
        <div id='sendPane'>
          <div id='sendAddress'>
            <input
              autoFocus
              type='text'
              placeholder='Where to send the asset (address)'
              value={sendAddress}
              onChange={(e) => this.setState({ sendAddress: e.target.value })}
            />
          </div>
          <div id='sendAmount'>
            <input
              type='text'
              value={sendAmount}
              placeholder='Amount'
              onChange={(e) => this.setState({ sendAmount: e.target.value })}
            />
          </div>
          <button id='sendAsset' data-tip data-for='assetTip' onClick={() => this.nextAsset()}>{selectedAsset}</button>
          <ReactTooltip class='solidTip' id='assetTip' place='bottom' type='dark' effect='solid'>
            <span>Toggle NEO / GAS / Assets</span>
          </ReactTooltip>
          <button id='doSend' onClick={this.openAndValidate}>Send Asset</button>
        </div>
        <div id='confirmPane' onClick={this.sendTransaction}>
          <button ref={node => { this.confirmButton = node }}>Confirm Transaction</button>
        </div>
      </SplitPane>
    )
  }
}
