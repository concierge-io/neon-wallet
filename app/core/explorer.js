// @flow
import { openExternal } from './electron'
import { NETWORK, EXPLORER } from '../core/constants'

export const getExplorerLink = (net: NetworkType, explorer: ExplorerType, txid: string) => {
  let base
  if (explorer === EXPLORER.NEO_TRACKER) {
    if (net === NETWORK.MAIN) {
      base = 'https://neotracker.io/tx/'
    } else {
      base = 'https://testnet.neotracker.io/tx/'
    }
  } else if (explorer === EXPLORER.NEO_SCAN) {
    if (net === NETWORK.MAIN) {
      base = 'https://neoscan.io/transaction/'
    } else {
      base = 'https://neoscan-testnet.io/transaction/'
    }
  } else {
    if (net === NETWORK.MAIN) {
      base = 'http://antcha.in/tx/hash/'
    } else {
      base = 'http://testnet.antcha.in/tx/hash/'
    }
  }
  return `${base}${txid}`
}

export const openExplorer = (net: NetworkType, explorer: ExplorerType, txid: string) =>
  openExternal(getExplorerLink(net, explorer, txid))
