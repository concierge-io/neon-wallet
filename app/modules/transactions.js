// @flow
/* eslint-disable camelcase */

import { ASSETS_LABELS, ASSETS, ASSET_ID } from '../core/constants'
import { validateTransactionBeforeSending } from '../core/wallet'
import { getTransactionHistory, doSendAsset, hardwareDoSendAsset, getRPCEndpoint, getBalance, serializeTransaction } from 'neon-js'
import {setTransactionHistory, getNeo, getGas, createCoin} from './wallet'
import { log } from '../util/Logs'
import { showErrorNotification, showInfoNotification, showSuccessNotification } from './notifications'
import { getWif, getPublicKey, getSigningFunction, getAddress, LOGOUT } from './account'
import { getNetwork } from './metadata'
import asyncWrap from '../core/asyncHelper'
import { buildScript } from '../sc/scriptBuilder'
import { reverseHex } from 'neon-js/src/utils'
import { doInvokeScript } from 'neon-js/src/api'
import { getScriptHashFromAddress, getAccountFromWIFKey } from 'neon-js/src/wallet'
import { invocationTx } from 'neon-js/src/transactions/create'
import { signTransaction } from 'neon-js/src/transactions/index'
import { Query } from '../core/query'

// Constants
export const TOGGLE_ASSET = 'TOGGLE_ASSET'
export const LOADING_TRANSACTIONS = 'LOADING_TRANSACTIONS'

export const toggleAsset = ( symbol: string,  hash: string) => ({
  type: TOGGLE_ASSET,
  asset: symbol,
  hash: hash
})

export const setIsLoadingTransaction = (isLoading: boolean) => ({
  type: LOADING_TRANSACTIONS,
  payload: {
    isLoadingTransactions: isLoading
  }
})

export const syncTransactionHistory = (net: NetworkType, address: string) => async (dispatch: DispatchType) => {
  dispatch(setIsLoadingTransaction(true))
  const [err, transactions] = await asyncWrap(getTransactionHistory(net, address))
  if (!err && transactions) {
    const txs = transactions.map(({ NEO, GAS, txid, block_index, neo_sent, neo_gas }: TransactionHistoryType) => ({
      type: neo_sent ? ASSETS.NEO : ASSETS.GAS,
      amount: neo_sent ? NEO : GAS,
      txid,
      block_index
    }))
    dispatch(setIsLoadingTransaction(false))
    dispatch(setTransactionHistory(txs))
  } else {
    dispatch(setIsLoadingTransaction(false))
  }
}

export const sendTransaction = (sendAddress: string, sendAmount: string) => async (dispatch: DispatchType, getState: GetStateType): Promise<*> => {

  const state = getState()
  const wif = getWif(state)
  const address = getAddress(state)
  const net = getNetwork(state)
  const neo = getNeo(state)
  const gas = getGas(state)
  const selectedAsset = getSelectedAsset(state)
  const selectedHash = getSelectedHash(state)
  const balances = state.nep.balances
  const signingFunction = getSigningFunction(state)
  const account = getAccountFromWIFKey(wif)
  const publicKey = account.publicKeyEncoded
  const privateKey = account.privateKey
  const tokens = state.nep.tokens

  const rejectTransaction = (message: string) => dispatch(showErrorNotification({ message }))
  const { error, valid } = validateTransactionBeforeSending(neo, gas, selectedAsset, selectedHash, balances, sendAddress, sendAmount)

  if (valid) {
    if(selectedHash){

      let parsedValue = Math.round(sendAmount * Math.pow(10, tokens[selectedHash].decimals))

      //run this logic for nep5 contracts
      doTransferToken(net, selectedHash.slice(2), wif, publicKey, privateKey, address, sendAddress, parsedValue, 0, signingFunction)
        .then((result) => {

          if(result.result){
            return dispatch(showSuccessNotification({ message: 'The transaction was successful. Wait until the transfer is approved to see it on the balance.' }))
          }else{
            return rejectTransaction('Transaction failed!')
          }
      })

    }else{

      const selfAddress = address
      const assetName = selectedAsset === ASSETS_LABELS.NEO ? ASSETS.NEO : ASSETS.GAS
      let sendAsset = {}
      sendAsset[assetName] = sendAmount

      dispatch(showInfoNotification({ message: 'Sending Transaction...', autoDismiss: 0 }))
      log(net, 'SEND', selfAddress, { to: sendAddress, asset: selectedAsset, amount: sendAmount })

      const isHardwareSend = !publicKey

      let sendAssetFn
      if (isHardwareSend) {
        dispatch(showInfoNotification({ message: 'Please sign the transaction on your hardware device', autoDismiss: 0 }))
        sendAssetFn = () => hardwareDoSendAsset(net, sendAddress, publicKey, sendAsset, signingFunction)
      } else {
        sendAssetFn = () => doSendAsset(net, sendAddress, wif, sendAsset)
      }

      const [err, response] = await asyncWrap(sendAssetFn())
      if (err || response.result === undefined || response.result === false) {
        return rejectTransaction('Transaction failed!')
      } else {
        return dispatch(showSuccessNotification({ message: 'Transaction complete! Your balance will automatically update when the blockchain has processed it.' }))
      }
    }

  } else {
    return rejectTransaction(error)
  }
}

/**
 * Transfers NEP5 Tokens.
 * @param {string} net
 * @param {string} scriptHash
 * @param {string} wif
 * @param {string} publickey
 * @param {string} fromAddress
 * @param {string} toAddress
 * @param {number} transferAmount
 * @param {number} gasCost
 * @param {function} signingFunction
 * @return {Promise<Response>} RPC response
 */
const doTransferToken = (net, scriptHash, wif, publickey, privatekey, fromAddress, toAddress, transferAmount, gasCost = 0, signingFunction = null) => {
  const rpcEndpointPromise = getRPCEndpoint(net)
  const balancePromise = getBalance(net, fromAddress)
  let signedTx
  let endpt
  return Promise.all([rpcEndpointPromise, balancePromise])
    .then((values) => {
      endpt = values[0]
      const balances = values[1]
      const fromAddrScriptHash = reverseHex(getScriptHashFromAddress(fromAddress))
      const intents = [
        { assetId: ASSET_ID.GAS, value: 0.00000001, scriptHash: fromAddrScriptHash }
      ]
      const toAddrScriptHash = reverseHex(getScriptHashFromAddress(toAddress))
      const invoke = { scriptHash, operation: 'transfer', args: [fromAddrScriptHash, toAddrScriptHash, transferAmount] }
      const unsignedTx = invocationTx(publickey, balances, intents, invoke, gasCost, { version: 1 })
      if (signingFunction) {
        return signingFunction(unsignedTx, publickey)
      } else {
        return signTransaction(unsignedTx, privatekey)
      }
    })
    .then((signedResult) => {
      signedTx = signedResult
      return sendRawTransaction(signedTx).execute(endpt)
    })
    .then((res) => {
      if (res.result === true) {
        res.txid = signedTx
      }
      return res
    })
}

export const sendRawTransaction = (transaction) => {
  const serialized = typeof (transaction) === 'object' ? serializeTransaction(transaction) : transaction
  return new Query({
    method: 'sendrawtransaction',
    params: [serialized]
  })
}

// state getters
export const getSelectedAsset = (state) => state.transactions.selectedAsset
export const getSelectedHash = (state) => state.transactions.selectedHash
export const getIsLoadingTransactions = (state) => state.transactions.isLoadingTransactions

const initialState = {
  selectedAsset: ASSETS_LABELS.NEO,
  selectedHash: null,
  isLoadingTransactions: false
}

export default (state: Object = initialState, action: Object) => {
  switch (action.type) {
    case LOADING_TRANSACTIONS:
      const { isLoadingTransactions } = action.payload
      return {
        ...state,
        isLoadingTransactions
      }
    case TOGGLE_ASSET:
      return {
        ...state,
        selectedAsset: action.asset,
        selectedHash: action.hash
      }
    case LOGOUT:
      return initialState
    default:
      return state
  }
}
