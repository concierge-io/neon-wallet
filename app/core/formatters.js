// @flow
import { truncateNumber } from './math'

const GAS_DECIMAL_LENGTH = 7
const GAS_DECIMAL_SHORT_DISPLAY_LENGTH = 4

const FIAT_DECIMAL_LENGTH = 2

export const formatGAS = (gas: number | string, shortDisplay: boolean = false): string => {
  const decimalLength = shortDisplay ? GAS_DECIMAL_SHORT_DISPLAY_LENGTH : GAS_DECIMAL_LENGTH
  return truncateNumber(parseFloat(gas), decimalLength).toFixed(decimalLength)
}

export const formatNEO = (neo: number | string): number => parseInt(neo, 10)

export const formatFiat = (value: number | string): string => parseFloat(value).toFixed(FIAT_DECIMAL_LENGTH)
