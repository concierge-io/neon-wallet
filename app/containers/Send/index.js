// @flow
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import Send from './Send'
import { sendTransaction, toggleAsset, getSelectedAsset, getSelectedHash } from '../../modules/transactions'
import { showErrorNotification } from '../../modules/notifications'
import { togglePane, getConfirmPane } from '../../modules/dashboard'
import { getNeo, getGas } from '../../modules/wallet'
import { getBalances, getTokens, getNep5 } from '../../modules/nep'


const mapStateToProps = (state) => ({
  neo: getNeo(state),
  gas: getGas(state),
  selectedAsset: getSelectedAsset(state),
  selectedHash: getSelectedHash(state),
  confirmPane: getConfirmPane(state),
  nep5: getNep5(state),
  tokens: getTokens(state),
  balances: getBalances(state)
})

const actionCreators = {
  sendTransaction,
  toggleAsset,
  togglePane,
  showErrorNotification
}

const mapDispatchToProps = dispatch => bindActionCreators(actionCreators, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(Send)
