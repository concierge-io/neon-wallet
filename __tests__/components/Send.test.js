import React from 'react'
import { Provider } from 'react-redux'
import thunk from 'redux-thunk'
import configureStore from 'redux-mock-store'
import { shallow, mount } from 'enzyme'
import { TOGGLE_SEND_PANE } from '../../app/modules/dashboard'
import { TOGGLE_ASSET } from '../../app/modules/transactions'
import { SHOW_NOTIFICATION, HIDE_NOTIFICATIONS, DEFAULT_POSITION } from '../../app/modules/notifications'
import { NOTIFICATION_LEVELS } from '../../app/core/constants'
import Send from '../../app/containers/Send'
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'

jest.useFakeTimers()

const axiosMock = new MockAdapter(axios)
axiosMock.onAny().reply(200)

const initialState = {
  account: {
    wif: 'L4SLRcPgqNMAMwM3nFSxnh36f1v5omjPg3Ewy1tg2PnEon8AcHou',
    address: 'AWy7RNBVr9vDadRMK9p7i7Z1tL7GrLAxoh'
  },
  metadata: {
    network: 'TestNet'
  },
  wallet: {
    Neo: 5,
    Gas: 1
  },
  transactions: {
    selectedAsset: 'Neo'
  },
  dashboard: {
    confirmPane: true
  }
}

const setup = (state = initialState, shallowRender = true) => {
  const store = configureStore([thunk])(state)

  let wrapper
  if (shallowRender) {
    wrapper = shallow(<Send store={store} />)
  } else {
    wrapper = mount(
      <Provider store={store}>
        <Send />
      </Provider>
    )
  }

  return {
    store,
    wrapper
  }
}

const changeAmount = (amount, wrapper) => {
  const amountField = wrapper.find('input[placeholder="Amount"]')
  amountField.instance().value = amount
  amountField.simulate('change')
}

const changeAddress = (address, wrapper) => {
  const addressField = wrapper.find('input[placeholder="Where to send the asset (address)"]')
  addressField.instance().value = address
  addressField.simulate('change')
}

describe('Send', () => {
  test('renders without crashing', (done) => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
    done()
  })

  test('toggleAsset button is getting called on click', (done) => {
    const { wrapper, store } = setup()
    const deepWrapper = wrapper.dive()

    deepWrapper.find('#sendAsset').simulate('click')

    const actions = store.getActions()
    expect(actions.length === 1).toEqual(true)
    expect(actions[0].type === TOGGLE_ASSET).toEqual(true)
    done()
  })

  test('sendAsset button is getting called correctly with no address or amount sent', (done) => {
    const { wrapper, store } = setup(initialState, false)
    wrapper.find('#doSend').simulate('click')

    jest.runAllTimers()
    const actions = store.getActions()
    expect(actions.length).toEqual(2)
    expect(actions[0]).toEqual({
      type: HIDE_NOTIFICATIONS,
      payload: {
        dismissible: true,
        position: DEFAULT_POSITION
      }
    })
    expect(actions[1]).toEqual({
      type: SHOW_NOTIFICATION,
      payload: expect.objectContaining({
        message: 'Please specify an address and amount',
        level: NOTIFICATION_LEVELS.ERROR
      })
    })
    done()
  })

  test('sendAsset button is getting called correctly with invalid address', (done) => {
    const { wrapper, store } = setup(initialState, false)

    changeAddress('This-is-an-invalid-address', wrapper)
    changeAmount(initialState.wallet.Neo, wrapper)

    wrapper.find('#doSend').simulate('click')

    jest.runAllTimers()
    const actions = store.getActions()
    expect(actions.length).toEqual(2)
    expect(actions[0]).toEqual({
      type: HIDE_NOTIFICATIONS,
      payload: {
        dismissible: true,
        position: DEFAULT_POSITION
      }
    })
    expect(actions[1]).toEqual({
      type: SHOW_NOTIFICATION,
      payload: expect.objectContaining({
        message: 'The address you entered was not valid.',
        level: NOTIFICATION_LEVELS.ERROR
      })
    })
    done()
  })

  test('sendAsset button is getting called correctly with fractional NEO', (done) => {
    const { wrapper, store } = setup(initialState, false)

    changeAddress(initialState.account.address, wrapper)
    changeAmount(initialState.wallet.Neo + 0.5, wrapper)

    wrapper.find('#doSend').simulate('click')

    jest.runAllTimers()
    const actions = store.getActions()
    expect(actions.length).toEqual(2)
    expect(actions[0]).toEqual({
      type: HIDE_NOTIFICATIONS,
      payload: {
        dismissible: true,
        position: DEFAULT_POSITION
      }
    })
    expect(actions[1]).toEqual({
      type: SHOW_NOTIFICATION,
      payload: expect.objectContaining({
        message: 'You cannot send fractional amounts of Neo.',
        level: NOTIFICATION_LEVELS.ERROR
      })
    })
    done()
  })

  test('sendAsset button is getting called correctly with no NEO amount', (done) => {
    const { wrapper, store } = setup(initialState, false)

    changeAddress(initialState.account.address, wrapper)

    wrapper.find('#doSend').simulate('click')

    jest.runAllTimers()
    const actions = store.getActions()
    expect(actions.length).toEqual(2)
    expect(actions[0]).toEqual({
      type: HIDE_NOTIFICATIONS,
      payload: expect.objectContaining({
        dismissible: true,
        position: DEFAULT_POSITION
      })
    })
    expect(actions[1]).toEqual({
      type: SHOW_NOTIFICATION,
      payload: expect.objectContaining({
        message: 'Please specify an address and amount',
        level: NOTIFICATION_LEVELS.ERROR
      })
    })
    done()
  })

  test('sendAsset button is getting called correctly with not enough NEO', (done) => {
    const { wrapper, store } = setup(initialState, false)

    changeAddress(initialState.account.address, wrapper)
    changeAmount(initialState.wallet.Neo + 1, wrapper)

    wrapper.find('#doSend').simulate('click')

    jest.runAllTimers()
    const actions = store.getActions()
    expect(actions.length).toEqual(2)
    expect(actions[0]).toEqual({
      type: HIDE_NOTIFICATIONS,
      payload: {
        dismissible: true,
        position: DEFAULT_POSITION
      }
    })
    expect(actions[1]).toEqual({
      type: SHOW_NOTIFICATION,
      payload: expect.objectContaining({
        message: 'You do not have enough NEO to send.',
        level: NOTIFICATION_LEVELS.ERROR
      })
    })
    done()
  })

  test('sendAsset button is getting called correctly with negative NEO', (done) => {
    const { wrapper, store } = setup(initialState, false)

    changeAddress(initialState.account.address, wrapper)
    changeAmount(-1, wrapper)

    wrapper.find('#doSend').simulate('click')

    jest.runAllTimers()
    const actions = store.getActions()
    expect(actions.length).toEqual(2)
    expect(actions[0]).toEqual({
      type: HIDE_NOTIFICATIONS,
      payload: {
        dismissible: true,
        position: DEFAULT_POSITION
      }
    })
    expect(actions[1]).toEqual({
      type: SHOW_NOTIFICATION,
      payload: expect.objectContaining({
        message: 'You cannot send negative amounts of an asset.',
        level: NOTIFICATION_LEVELS.ERROR
      })
    })
    done()
  })

  test('sendAsset button is getting called correctly with correct NEO amount', (done) => {
    const { wrapper, store } = setup(initialState, false)

    changeAddress(initialState.account.address, wrapper)
    changeAmount(initialState.wallet.Neo - 1, wrapper)

    wrapper.find('#doSend').simulate('click')

    const actions = store.getActions()
    expect(actions.length).toEqual(1)
    expect(actions[0]).toEqual({
      type: TOGGLE_SEND_PANE,
      payload: expect.objectContaining({
        pane: 'confirmPane'
      })
    })
    done()
  })

  test('sendAsset button is getting called correctly for without enough Gas', (done) => {
    const gasState = Object.assign({}, initialState, { transactions: { selectedAsset: 'Gas' } })
    const { wrapper, store } = setup(gasState, false)

    changeAddress(initialState.account.address, wrapper)
    changeAmount(initialState.wallet.Gas + 1, wrapper)

    wrapper.find('#doSend').simulate('click')

    jest.runAllTimers()
    const actions = store.getActions()
    expect(actions.length).toEqual(2)
    expect(actions[0]).toEqual({
      type: HIDE_NOTIFICATIONS,
      payload: {
        dismissible: true,
        position: DEFAULT_POSITION
      }
    })
    expect(actions[1]).toEqual({
      type: SHOW_NOTIFICATION,
      payload: expect.objectContaining({
        message: 'You do not have enough GAS to send.',
        level: NOTIFICATION_LEVELS.ERROR
      })
    })
    done()
  })

  test('sendAsset button is getting called correctly with correct Gas amount', (done) => {
    const gasState = Object.assign({}, initialState, { transactions: { selectedAsset: 'Gas' } })
    const { wrapper, store } = setup(gasState, false)

    changeAddress(initialState.account.address, wrapper)
    changeAmount(initialState.wallet.Gas - 0.5, wrapper)

    wrapper.find('#doSend').simulate('click')

    const actions1 = store.getActions()
    expect(actions1.length).toEqual(1)
    expect(actions1[0]).toEqual({
      type: TOGGLE_SEND_PANE,
      payload: expect.objectContaining({
        pane: 'confirmPane'
      })
    })
    done()
  })

  test('send transaction is working correctly', (done) => {
    const { wrapper, store } = setup(initialState, false)

    changeAddress(initialState.account.address, wrapper)
    changeAmount(initialState.wallet.Neo - 1, wrapper)

    wrapper.find('#confirmPane').simulate('click')

    Promise.resolve('pause').then().then().then(() => {
      jest.runAllTimers()
      const actions = store.getActions()
      expect(actions.length).toEqual(4)
      expect(actions[0]).toEqual({
        type: HIDE_NOTIFICATIONS,
        payload: {
          dismissible: true,
          position: DEFAULT_POSITION
        }
      })
      expect(actions[1]).toEqual({
        type: SHOW_NOTIFICATION,
        payload: expect.objectContaining({
          message: 'Sending Transaction...',
          level: NOTIFICATION_LEVELS.INFO
        })
      })
      expect(actions[2]).toEqual({
        type: HIDE_NOTIFICATIONS,
        payload: {
          dismissible: true,
          position: DEFAULT_POSITION
        }
      })
      expect(actions[3]).toEqual({
        type: SHOW_NOTIFICATION,
        payload: expect.objectContaining({
          message: 'Transaction complete! Your balance will automatically update when the blockchain has processed it.',
          level: NOTIFICATION_LEVELS.SUCCESS
        })
      })
      done()
    }).catch(e => done.fail(e))
  })

  test('component is rendering correctly', (done) => {
    const { wrapper } = setup()
    const deepWrapper = wrapper.dive()

    const checkElements = [
      {
        element: deepWrapper.find('#sendAsset'),
        text: 'Neo'
      },
      {
        element: deepWrapper.find('#sendAmount'),
        text: ''
      },
      {
        element: deepWrapper.find('#sendAddress'),
        text: ''
      },
      {
        element: deepWrapper.find('#doSend'),
        text: 'Send Asset'
      },
      {
        element: deepWrapper.find('#confirmPane'),
        text: 'Confirm Transaction'
      }
    ]

    checkElements.forEach(item => {
      expect(item.element.length).toEqual(1)
      expect(item.element.text()).toEqual(item.text)
    })
    done()
  })
})
