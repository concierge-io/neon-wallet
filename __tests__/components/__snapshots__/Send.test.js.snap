// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send renders without crashing 1`] = `
<Send
  confirmPane={true}
  gas={1}
  neo={5}
  selectedAsset="Neo"
  sendTransaction={[Function]}
  showErrorNotification={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  toggleAsset={[Function]}
  togglePane={[Function]}
/>
`;
